### ISSUE TITLE
🐛 [firebase_core_web] Breaking change from 2.0.2 and up for Flutter web

### ISSUE THREAD

**Issue by Gorter-dev** [firebase/flutterfire#10228]
## Bug report

**Describe the bug**
Breaking change in Firebase_core_web from version 2.0.2 and up

### Steps to reproduce
Somewhere in main.dart: 
```
Firebase.initializeApp(
    name: '',
    options:  FirebaseOptions(
      apiKey: '',
      authDomain: '',
      databaseURL: '',
      projectId: '',
      storageBucket: '',
      messagingSenderId: '',
      appId: '',
      measurementId: '',
    ),
  ); 
```
(Items are filled in the app obviously)

Start the app and the error will be:
```
Error: [core/no-options] Firebase: Need to provide options, when not being deployed to hosting via source..
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 266:49                                  throw_
packages/firebase_core_web/src/firebase_core_web.dart 305:7                                                   app
packages/firebase_core/src/firebase.dart 56:41                                                                app
packages/firebase_auth_platform_interface/src/platform_interface/platform_interface_firebase_auth.dart 42:23  get app
packages/firebase_auth_web/firebase_auth_web.dart 119:68                                                      get delegate
packages/firebase_auth_web/firebase_auth_web.dart 93:47                                                       <fn>
dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 84:54                                            runBody
dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 123:5                                            _async
packages/firebase_auth_web/firebase_auth_web.dart 92:45                                                       <fn>
packages/firebase_core_web/src/firebase_core_web.dart 283:41                                                  <fn>
dart-sdk/lib/internal/iterable.dart 391:20                                                                    moveNext
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/operations.dart 748:20                              next
dart-sdk/lib/async/future.dart 520:16                                                                         wait
packages/firebase_core_web/src/firebase_core_web.dart 275:18                                                  initializeApp
dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 45:50                                            <fn>
dart-sdk/lib/async/zone.dart 1653:54                                                                          runUnary
dart-sdk/lib/async/future_impl.dart 147:18                                                                    handleValue
dart-sdk/lib/async/future_impl.dart 766:44                                                                    handleValueCallback
dart-sdk/lib/async/future_impl.dart 795:13                                                                    _propagateToListeners
dart-sdk/lib/async/future_impl.dart 566:5                                                                     [_completeWithValue]
dart-sdk/lib/async/future.dart 528:22                                                                         <fn>
dart-sdk/lib/async/zone.dart 1653:54                                                                          runUnary
dart-sdk/lib/async/future_impl.dart 147:18                                                                    handleValue
dart-sdk/lib/async/future_impl.dart 766:44                                                                    handleValueCallback
dart-sdk/lib/async/future_impl.dart 795:13                                                                    _propagateToListeners
dart-sdk/lib/async/future_impl.dart 566:5                                                                     [_completeWithValue]
dart-sdk/lib/async/future_impl.dart 639:7                                                                     callback
dart-sdk/lib/async/schedule_microtask.dart 40:11                                                              _microtaskLoop
dart-sdk/lib/async/schedule_microtask.dart 49:5                                                               _startMicrotaskLoop
dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 166:15                                           <fn>
```

### Expected behavior

App running like it does with version 2.0.1 

## Additional context

Looks to me like it's something in this commit: https://github.com/firebase/flutterfire/pull/10028

**Comment by darshankawar** [comment]
Thanks for the report @Gorter-dev 
I am unable to replicate this using the plugin example and `firebase_core_web: ^2.1.0`. I used the same below code snippet inside `InitializeApp` method that you provided and filled it with my local information.

```
Firebase.initializeApp(
    name: '',
    options:  FirebaseOptions(
      apiKey: '',
      authDomain: '',
      databaseURL: '',
      projectId: '',
      storageBucket: '',
      messagingSenderId: '',
      appId: '',
      measurementId: '',
    ),
  ); 
```

Running above in web compiled the app properly and then after clicking on `initialize default app`, it triggered below error which I think is expected:

`Error: [core/bad-app-name] Firebase: Illegal App name: '.`

I am on latest master channel. Can you provide exact steps to replicate along with `flutter doctor -v` and `pubspec.yaml` ? Also please take a look at [plugin example](https://github.com/firebase/flutterfire/tree/master/packages/firebase_core/firebase_core/example) and see if using it, you still get same error or not.

**Comment by Gorter-dev** [comment]
I just ran the example in FlutterFire - FirebaseCore/Example. This works just fine!

I narrowed down the issue to the following:
Our application works just fine when I remove the name parameter from the initialization call. However when I add this I get the error message from above.

Looks to me like somewhere in firebase_auth_web.dart it defaults to the default app.

**Comment by darshankawar** [comment]
> Our application works just fine when I remove the name parameter from the initialization call. However when I add this I get the error message from above.

This looks similar to https://github.com/firebase/flutterfire/issues/9400 which was reported mentioning an error is thrown on web if `name` is used, but since has been fixed. Please take a look at it and see if it helps in your case.

**Comment by bengarrison** [comment]
I just got this error as well.  Confirmed, the error won't throw if the named parameter is removed.

**Comment by darshankawar** [comment]
@bengarrison 
Can you provide a minimal code sample that shows the error along with `pubspec.yaml` ?

**Comment by Gorter-dev** [comment]
I think it has to do with this getInstance call in the commit I mentioned earlier.

![Schermafbeelding 2023-01-26 om 09 45 41](https://user-images.githubusercontent.com/52274327/214793098-a195cdc2-c4b1-4e18-bf29-a1424e300bac.png)

The problem itself seems easy enough, when a name parameter is added on the said version. The app won't start because of the feature: ```currentUser is now populated right at the start of the application without needing to wait for authStateChange```, because this works only for the DEFAULT_APP and not any named instance.

**Comment by Gorter-dev** [comment]
When I comment out this line:
```await FirebaseAuthWeb.instance.delegate.onWaitInitState();```

The error is gone.

![Schermafbeelding 2023-01-26 om 10 28 24](https://user-images.githubusercontent.com/52274327/214801611-9f7fe667-ac55-451a-a4e8-25fa430f1d65.png)

**Comment by Gorter-dev** [comment]
I figured the whole thing out:

When using:
firebase_auth_web 5.2.3 
firebase_core_web: 2.0.2
firebase_auth_platform_interface: 6.11.6

And having a not empty name in the firebase intializeApp function.

This static function get called initially:
![Schermafbeelding 2023-01-26 om 11 07 00](https://user-images.githubusercontent.com/52274327/214809391-76ef8fa8-abab-4419-ac13-92b5a1d426f2.png)

In this function there is a call to a getter named delegate
![Schermafbeelding 2023-01-26 om 11 09 16](https://user-images.githubusercontent.com/52274327/214809831-7da5cebf-6e4a-483c-b591-6fa2c380bd69.png)

this delegate uses app, app comes from the firebase_auth_platform_interface
![Schermafbeelding 2023-01-26 om 11 11 13](https://user-images.githubusercontent.com/52274327/214810257-d6c8dfa3-8eb3-4549-a1a5-d68b37941454.png)

This getter of app return the appInstance or ```Firebase.app()```

And due to the moment of calling the first static function the appInstance will always be null. Therefore returning the firebase.app(), whilst the project intializes a named app.

The solution (I think), will something delaying the call ```FirebaseAuthWeb.instance.delegate.onWaitInitState();``` to a later moment after initialization is completed, thus appInstance will not be null. 

I don't know what complications this will have for the feature: ```currentUser is now populated right at the start of the application without needing to wait for authStateChange```, this will therefore be also delayed.

Kind regards,

**Comment by darshankawar** [comment]
Thanks for the update and digging in @Gorter-dev 

With the details provided above, I was able to replicate the original error described using plugin example.

<img width="407" alt="Screenshot 2023-01-27 at 4 13 03 PM" src="https://user-images.githubusercontent.com/67046386/215067376-f638d62f-3dee-4f17-a9d7-4c88eae49346.png">

Added `name` parameter to `initializeApp` as below:

```
Future<void> initializeDefault() async {
    FirebaseApp app = await Firebase.initializeApp(
      name: 'test',
      options: FirebaseOptions(
```

Upon running the app which compiles properly but then tapping to initialize either the default or secondary app throw below exception:

<details>
<summary> console error log </summary>

```

Error: [core/no-options] Firebase: Need to provide options, when not being deployed to hosting via source..
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 288:49                                  throw_
packages/firebase_core_web/src/firebase_core_web.dart 305:7                                                   app
packages/firebase_core/src/firebase.dart 56:41                                                                app
packages/firebase_auth_platform_interface/src/platform_interface/platform_interface_firebase_auth.dart 42:23  get app
packages/firebase_auth_web/firebase_auth_web.dart 119:68                                                      get delegate
packages/firebase_auth_web/firebase_auth_web.dart 93:47                                                       <fn>
dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 84:54                                            runBody
dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 123:5                                            _async
packages/firebase_auth_web/firebase_auth_web.dart 92:45                                                       <fn>
packages/firebase_core_web/src/firebase_core_web.dart 283:41                                                  <fn>
dart-sdk/lib/internal/iterable.dart 393:20                                                                    moveNext
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/operations.dart 831:20                              next
dart-sdk/lib/async/future.dart 520:16                                                                         wait
packages/firebase_core_web/src/firebase_core_web.dart 275:18                                                  initializeApp
dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 45:50                                            <fn>
dart-sdk/lib/async/zone.dart 1660:54                                                                          runUnary
dart-sdk/lib/async/future_impl.dart 147:18                                                                    handleValue
dart-sdk/lib/async/future_impl.dart 767:44                                                                    handleValueCallback
dart-sdk/lib/async/future_impl.dart 796:13                                                                    _propagateToListeners
dart-sdk/lib/async/future_impl.dart 567:5                                                                     [_completeWithValue]
dart-sdk/lib/async/future.dart 528:22                                                                         <fn>
dart-sdk/lib/async/zone.dart 1660:54                                                                          runUnary
dart-sdk/lib/async/future_impl.dart 147:18                                                                    handleValue
dart-sdk/lib/async/future_impl.dart 767:44                                                                    handleValueCallback
dart-sdk/lib/async/future_impl.dart 796:13                                                                    _propagateToListeners
dart-sdk/lib/async/future_impl.dart 567:5                                                                     [_completeWithValue]
dart-sdk/lib/async/future_impl.dart 640:7                                                                     callback
dart-sdk/lib/async/schedule_microtask.dart 40:11                                                              _microtaskLoop
dart-sdk/lib/async/schedule_microtask.dart 49:5                                                               _startMicrotaskLoop
dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 177:15                                           <fn>

```

</details>

Removing or commenting out the `name` parameter doesn't throw the error and all functionalities work properly from plugin example.

**Comment by Gorter-dev** [comment]
Correct! for now just removing the name parameter will do. I do think that the error itself and the ```onWaitInitState``` should be called in a later moment to actually fix the problem

**Comment by Lyokone** [comment]
@oantajames Thanks for raising this issue.
You can use this hotfix in your `pubspec.yaml` to remove this behavior:
```
  firebase_auth_web:
    git:
      url: https://github.com/firebase/flutterfire.git
      ref: hotfix/web-user-null
      path: packages/firebase_auth/firebase_auth_web
```

We are currently waiting for some input from Google to fix the issue #10200 properly

**Comment by Gorter-dev** [comment]
@Lyokone, isn't delaying ```onWaitInitState``` a feasible option. Thus also not calling it from a static method

**Comment by Lyokone** [comment]
Have you updated to `firebase_auth: ^4.2.10`? It should solve this issue.

**Comment by Aburafik** [comment]
When running on android or ios, the name tag is required. But with flutter web the name has to be removed. Have been facing this issue for the past two weeks simple because my project is running on this three platforms. I just got it fixed after going through the answers provided above.

**Comment by russellwheatley** [comment]
Closing out as this has been fixed a while back in firebase_auth => 4.2.10.
